reconcilers/reconcilers.go

```diff
@@ -1273,6 +1274,413 @@ func (r *ChildReconciler[T, CT, CLT]) ourChild(resource T, obj CT) bool {
+		return nil, utilErrs.NewAggregate(childKeyErrors)
+	}
+	return childMap, nil
+}
+
+func (r *ChildSetReconciler[T, CT, CLT]) getValueOrDefault(m map[string]CT, key string, defaultValue CT) CT {
+	if val, ok := m[key]; ok {
+		return val
+	}
+	return defaultValue
+}
+
+func (r *ChildSetReconciler[T, CT, CLT]) reconcile(ctx context.Context, resource T) ([]CT, utilErrs.Aggregate) {
+	log := logr.FromContextOrDiscard(ctx)
+	c := RetrieveConfigOrDie(ctx)
+
+	children := r.ChildListType.DeepCopyObject().(CLT)
+	if err := c.List(ctx, children, r.listOptions(ctx, resource)...); err != nil {
+		return nil, utilErrs.NewAggregate([]error{err})
+	}
+	actualChildren := r.filterChildren(resource, children)
+	desiredChildren, err := r.desiredChildSet(ctx, resource)
+	if err != nil {
+		if errors.Is(err, OnlyReconcileChildStatus) {
+			return actualChildren, nil
+		}
+		return nil, utilErrs.NewAggregate([]error{err})
+	}
+	if internal.IsNil(desiredChildren) {
+		desiredChildren = []CT{}
+
+		if !r.hasChild(actualChildren) {
+			err = ClearFinalizer(ctx, resource, r.Finalizer)
+			return nil, utilErrs.NewAggregate([]error{err})
+		}
+	} else {
+		if err = AddFinalizer(ctx, resource, r.Finalizer); err != nil {
+			return nil, utilErrs.NewAggregate([]error{err})
+		}
+	}
+
+	actualChildrenMap, err1 := r.childListToMap(actualChildren)
+	if err1 != nil {
+		return nil, err1
+	}
+
+	desiredChildrenMap, err2 := r.childListToMap(desiredChildren)
+	if err2 != nil {
+		return nil, err2
+	}
+
+	emtpyCT := r.ChildType.DeepCopyObject().(CT)
+	var resultList []CT
+	var errorList []error
+	for _, desiredChild := range desiredChildren {
+		key, err3 := r.ChildKey(desiredChild)
+		if err3 != nil {
+			errorList = append(errorList, err3)
+		}
+
+		if !r.SkipOwnerReference {
+			if err = ctrl.SetControllerReference(resource, desiredChild, c.Scheme()); err != nil {
+				errorList = append(errorList, err)
+				continue
+			}
+		}
+
+		if !r.ourChild(resource, desiredChild) {
+			log.Info("One child returned from DesiredChildSet does not match OurChild, this can result in orphaned children", "child", namespaceName(desiredChild))
+		}
+
+		var result CT
+		actualChild := r.getValueOrDefault(actualChildrenMap, key, emtpyCT)
+		result, err = r.stamp.Manage(ctx, resource, actualChild, desiredChild)
+		if err != nil {
+			errorList = append(errorList, err)
+			continue
+		}
+		resultList = append(resultList, result)
+	}
+
+	// Undesired keys
+	for _, actualChild := range actualChildren {
+		key, err3 := r.ChildKey(actualChild)
+		if err3 != nil {
+			errorList = append(errorList, err3)
+		}
+		if _, ok := desiredChildrenMap[key]; !ok {
+			_, err = r.stamp.Manage(ctx, resource, actualChild, emtpyCT)
+			if err != nil {
+				errorList = append(errorList, err)
+			}
+		}
+	}
+
+	return resultList, utilErrs.NewAggregate(errorList)
+}
+
+func (r *ChildSetReconciler[T, CT, CLT]) hasChild(childSet []CT) bool {
+	hasChild := false
+	if !internal.IsNil(childSet) {
```

Comment:
checking the length is a more common way to nil check a slice, unless nil and length 0 have different semantic values.